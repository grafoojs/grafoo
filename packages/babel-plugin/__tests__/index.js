import pluginTester from "babel-plugin-tester";
import plugin from "../src";

pluginTester({
  plugin,
  pluginName: "@grafoo/babel-plugin",
  pluginOptions: {
    schema: "__tests__/schema.graphql",
    idFields: ["id"]
  },
  tests: {
    "should throw if a import is not default": {
      code: 'import { gql } from "@grafoo/core/tag";',
      error: true
    },
    "should throw if a schema is not present on the root directory": {
      pluginOptions: {
        idFields: ["id"]
      },
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`{ hello }\`;
      `,
      error: true
    },
    "should throw if a tagged template string literal has expressions in it": {
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`{ user(id: "\${1}") { name } }\`;
      `,
      error: true
    },
    "should remove the imported path": {
      code: 'import gql from "@grafoo/core/tag";',
      snapshot: true
    },
    "should throw if idFields is not defined": {
      pluginOptions: {
        schema: "__tests__/schema.graphql"
      },
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`{ hello }\`;
      `,
      error: true
    },
    "should throw if during client instatiation options is passed with a type other then object": {
      code: `
        import createClient from "@grafoo/core";
        let query = createClient(someTransport, "I AM ERROR");
      `,
      error: true
    },
    "should throw if the type of some field in `idFields` is not of type string": {
      pluginOptions: {
        schema: "__tests__/schema.graphql",
        idFields: ["id", true]
      },
      code: `
        import createClient from "@grafoo/core";
        let query = createClient(someTransport);
      `,
      error: true
    },
    "should replace a tagged template literal with the compiled grafoo object": {
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`
          query($start: Int!, $offset: Int!, $id: ID!) {
            posts(start: $start, offset: $offset) {
              title
              body
              createdAt
              tags { name }
              authors { name username }
            }
            user(id: $id) { name username }
          }
        \`;
      `,
      snapshot: true
    },
    "should compress the query string if the option compress is specified": {
      pluginOptions: {
        schema: "__tests__/schema.graphql",
        idFields: ["id"],
        compress: true
      },
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`
          query($start: Int!, $offset: Int!, $id: ID!) {
            posts(start: $start, offset: $offset) {
              title
              body
              createdAt
              tags { name }
              authors { name username }
            }
            user(id: $id) { name username }
          }
        \`;
      `,
      snapshot: true
    },
    "should generate md5 hash and add it to object if the option generateIds is specified": {
      pluginOptions: {
        schema: "__tests__/schema.graphql",
        idFields: ["id"],
        generateIds: true
      },
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`
          query($start: Int!, $offset: Int!, $id: ID!) {
            posts(start: $start, offset: $offset) {
              title
              body
              createdAt
              tags { name }
              authors { name username }
            }
            user(id: $id) { name username }
          }
        \`;
      `,
      snapshot: true
    },
    "should not generate md5 hash and add it to object if the option generateIds is falsey": {
      pluginOptions: {
        schema: "__tests__/schema.graphql",
        idFields: ["id"]
      },
      code: `
        import gql from "@grafoo/core/tag";
        let query = gql\`
          query($start: Int!, $offset: Int!, $id: ID!) {
            posts(start: $start, offset: $offset) {
              title
              body
              createdAt
              tags { name }
              authors { name username }
            }
            user(id: $id) { name username }
          }
        \`;
      `,
      snapshot: true
    },
    "should include `idFields` in the client instantiation if options are not provided": {
      code: `
        import createClient from "@grafoo/core";
        let query = createClient(someTransport);
      `,
      snapshot: true
    },
    "should include `idFields` in the client instantiation if not present in options": {
      code: `
        import createClient from "@grafoo/core";
        let query = createClient(someTransport, {});
      `,
      snapshot: true
    },
    "should include `idFields` in the client instantiation if options is a variable": {
      code: `
        import createClient from "@grafoo/core";
        let options = {};
        let query = createClient(someTransport, options);
      `,
      snapshot: true
    },
    "should overide `idFields` in the client instantiation if options is a variable": {
      code: `
        import createClient from "@grafoo/core";
        let options = { idFields: ["err"] };
        let query = createClient(someTransport, options);
      `,
      snapshot: true
    },
    "should throw if `idFields` in the client instantiation if options is not an object variable": {
      code: `
        import createClient from "@grafoo/core";
        let options = [];
        let query = createClient(someTransport, options);
      `,
      error: true
    },
    "should include `idFields` in the client instantiation even if options are provided": {
      code: `
        import createClient from "@grafoo/core";
        let query = createClient(someTransport, {
          headers: () => ({ authorization: "some-token" })
        });
      `,
      snapshot: true
    }
  }
});
