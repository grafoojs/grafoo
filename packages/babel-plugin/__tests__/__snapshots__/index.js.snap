// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@grafoo/babel-plugin should compress the query string if the option compress is specified: should compress the query string if the option compress is specified 1`] = `

import gql from "@grafoo/core/tag";
let query = gql\`
  query($start: Int!, $offset: Int!, $id: ID!) {
    posts(start: $start, offset: $offset) {
      title
      body
      createdAt
      tags { name }
      authors { name username }
    }
    user(id: $id) { name username }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

let query = {
  query:
    "query($id:ID!,$offset:Int!,$start:Int!){posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}user(id:$id){id name username}}",
  paths: {
    "posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}": {
      name: "posts",
      args: ["offset", "start"],
    },
    "user(id:$id){id name username}": {
      name: "user",
      args: ["id"],
    },
  },
};


`;

exports[`@grafoo/babel-plugin should generate md5 hash and add it to object if the option generateIds is specified: should generate md5 hash and add it to object if the option generateIds is specified 1`] = `

import gql from "@grafoo/core/tag";
let query = gql\`
  query($start: Int!, $offset: Int!, $id: ID!) {
    posts(start: $start, offset: $offset) {
      title
      body
      createdAt
      tags { name }
      authors { name username }
    }
    user(id: $id) { name username }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

let query = {
  id: "6e0697df8f2453f2643bbd1e8a39c348",
  query:
    "query ($id: ID!, $offset: Int!, $start: Int!) {\\n  posts(offset: $offset, start: $start) {\\n    authors {\\n      id\\n      name\\n      username\\n    }\\n    body\\n    createdAt\\n    id\\n    tags {\\n      id\\n      name\\n    }\\n    title\\n  }\\n  user(id: $id) {\\n    id\\n    name\\n    username\\n  }\\n}",
  paths: {
    "posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}": {
      name: "posts",
      args: ["offset", "start"],
    },
    "user(id:$id){id name username}": {
      name: "user",
      args: ["id"],
    },
  },
};


`;

exports[`@grafoo/babel-plugin should include \`idFields\` in the client instantiation even if options are provided: should include \`idFields\` in the client instantiation even if options are provided 1`] = `

import createClient from "@grafoo/core";
let query = createClient(someTransport, {
  headers: () => ({ authorization: "some-token" })
});

      ↓ ↓ ↓ ↓ ↓ ↓

import createClient from "@grafoo/core";
let query = createClient(someTransport, {
  headers: () => ({
    authorization: "some-token",
  }),
  idFields: ["id"],
});


`;

exports[`@grafoo/babel-plugin should include \`idFields\` in the client instantiation if not present in options: should include \`idFields\` in the client instantiation if not present in options 1`] = `

import createClient from "@grafoo/core";
let query = createClient(someTransport, {});

      ↓ ↓ ↓ ↓ ↓ ↓

import createClient from "@grafoo/core";
let query = createClient(someTransport, {
  idFields: ["id"],
});


`;

exports[`@grafoo/babel-plugin should include \`idFields\` in the client instantiation if options are not provided: should include \`idFields\` in the client instantiation if options are not provided 1`] = `

import createClient from "@grafoo/core";
let query = createClient(someTransport);

      ↓ ↓ ↓ ↓ ↓ ↓

import createClient from "@grafoo/core";
let query = createClient(someTransport, {
  idFields: ["id"],
});


`;

exports[`@grafoo/babel-plugin should include \`idFields\` in the client instantiation if options is a variable: should include \`idFields\` in the client instantiation if options is a variable 1`] = `

import createClient from "@grafoo/core";
let options = {};
let query = createClient(someTransport, options);

      ↓ ↓ ↓ ↓ ↓ ↓

import createClient from "@grafoo/core";
let options = {
  idFields: ["id"],
};
let query = createClient(someTransport, options);


`;

exports[`@grafoo/babel-plugin should not generate md5 hash and add it to object if the option generateIds is falsey: should not generate md5 hash and add it to object if the option generateIds is falsey 1`] = `

import gql from "@grafoo/core/tag";
let query = gql\`
  query($start: Int!, $offset: Int!, $id: ID!) {
    posts(start: $start, offset: $offset) {
      title
      body
      createdAt
      tags { name }
      authors { name username }
    }
    user(id: $id) { name username }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

let query = {
  query:
    "query ($id: ID!, $offset: Int!, $start: Int!) {\\n  posts(offset: $offset, start: $start) {\\n    authors {\\n      id\\n      name\\n      username\\n    }\\n    body\\n    createdAt\\n    id\\n    tags {\\n      id\\n      name\\n    }\\n    title\\n  }\\n  user(id: $id) {\\n    id\\n    name\\n    username\\n  }\\n}",
  paths: {
    "posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}": {
      name: "posts",
      args: ["offset", "start"],
    },
    "user(id:$id){id name username}": {
      name: "user",
      args: ["id"],
    },
  },
};


`;

exports[`@grafoo/babel-plugin should overide \`idFields\` in the client instantiation if options is a variable: should overide \`idFields\` in the client instantiation if options is a variable 1`] = `

import createClient from "@grafoo/core";
let options = { idFields: ["err"] };
let query = createClient(someTransport, options);

      ↓ ↓ ↓ ↓ ↓ ↓

import createClient from "@grafoo/core";
let options = {
  idFields: ["id"],
};
let query = createClient(someTransport, options);


`;

exports[`@grafoo/babel-plugin should remove the imported path: should remove the imported path 1`] = `

import gql from "@grafoo/core/tag";

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`@grafoo/babel-plugin should replace a tagged template literal with the compiled grafoo object: should replace a tagged template literal with the compiled grafoo object 1`] = `

import gql from "@grafoo/core/tag";
let query = gql\`
  query($start: Int!, $offset: Int!, $id: ID!) {
    posts(start: $start, offset: $offset) {
      title
      body
      createdAt
      tags { name }
      authors { name username }
    }
    user(id: $id) { name username }
  }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

let query = {
  query:
    "query ($id: ID!, $offset: Int!, $start: Int!) {\\n  posts(offset: $offset, start: $start) {\\n    authors {\\n      id\\n      name\\n      username\\n    }\\n    body\\n    createdAt\\n    id\\n    tags {\\n      id\\n      name\\n    }\\n    title\\n  }\\n  user(id: $id) {\\n    id\\n    name\\n    username\\n  }\\n}",
  paths: {
    "posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}": {
      name: "posts",
      args: ["offset", "start"],
    },
    "user(id:$id){id name username}": {
      name: "user",
      args: ["id"],
    },
  },
};


`;
